include/ginkgo/extensions/kokkos/spaces.hpp

```diff
@@ -0,0 +1,290 @@
+
+#ifndef GINKGO_EXTENSIONS_KOKKOS_SPACES_HPP
+#define GINKGO_EXTENSIONS_KOKKOS_SPACES_HPP
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_EXTENSION_KOKKOS
+
+
+#include <Kokkos_Core.hpp>
+
+
+#include <ginkgo/core/base/executor.hpp>
+
+
+namespace gko {
+namespace ext {
+namespace kokkos {
+namespace detail {
+
+
+/**
+ * Helper to check if an executor type can access the memory of an memory space
+ *
+ * @tparam MemorySpace  Type fulfilling the Kokkos MemorySpace concept.
+ * @tparam ExecType  One of the Ginkgo executor types.
+ */
+template <typename MemorySpace, typename ExecType>
+struct compatible_space
+    : std::integral_constant<bool, Kokkos::has_shared_space ||
+                                       Kokkos::has_shared_host_pinned_space> {};
+
+template <>
+struct compatible_space<Kokkos::HostSpace, ReferenceExecutor> : std::true_type {
+};
+
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, ReferenceExecutor> {
+    // need manual implementation of std::integral_constant because,
+    // while compiling for cuda, somehow bool is replaced by __nv_bool
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::HostSpace, MemorySpace>::accessible;
+};
+
+#ifdef KOKKOS_ENABLE_OPENMP
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, OmpExecutor>
+    : compatible_space<MemorySpace, ReferenceExecutor> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, CudaExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::Cuda, MemorySpace>::accessible;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, HipExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::HIP, MemorySpace>::accessible;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_SYCL
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, DpcppExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::Experimental::SYCL,
+                                   MemorySpace>::accessible;
+};
+#endif
+
+
+/**
+ * Checks if the memory space is accessible by the executor
+ *
+ * @tparam MemorySpace  A Kokkos memory space type
+ * @tparam ExecType  A Ginkgo executor type
+ *
+ * @return  true if the memory space is accessible by the executor
+ */
+template <typename MemorySpace, typename ExecType>
+inline bool check_compatibility(MemorySpace, std::shared_ptr<const ExecType>)
+{
+    return compatible_space<MemorySpace, ExecType>::value;
+}
+
+
+/**
+ * Checks if the memory space is accessible by the executor
+ *
+ * @tparam MemorySpace  A Kokkos memory space
+ * @param exec  The executor which should access the memory space
+ *
+ * @return  true if the memory space is accessible by the executor
+ */
+
+template <typename MemorySpace>
+inline bool check_compatibility(MemorySpace space,
```

Comment:
move the impl to later

Reply:
I don't think this rule applies to freestanding functions.